src/render/hw/draw/hw_dynamic_rrect_draw.cc

```diff
@@ -13,20 +14,73 @@ namespace skity {
 
 HWDynamicRRectDraw::HWDynamicRRectDraw(Matrix transform, RRect rrect,
                                        Paint paint)
-    : HWDynamicDraw(transform, paint.GetBlendMode()),
-      rrect_(std::move(rrect)),
-      paint_(std::move(paint)) {}
+    : HWDynamicDraw(transform, paint.GetBlendMode()) {
+  batch_group_.items.emplace_back(std::move(rrect));
+  batch_group_.paints.emplace_back(std::move(paint));
+  batch_group_.transforms.emplace_back(std::move(transform));
+}
 
-void HWDynamicRRectDraw::OnGenerateDrawStep(ArrayList<HWDrawStep *, 2> &steps,
-                                            HWDrawContext *context) {
-  auto arena_allocator = context->arena_allocator;
+bool HWDynamicRRectDraw::OnMergeIfPossible(HWDraw* draw) {
+  if (!HWDynamicDraw::OnMergeIfPossible(draw)) {
+    return false;
+  }
+  auto rrect_draw = static_cast<HWDynamicRRectDraw*>(draw);
+  auto& rrect_group = rrect_draw->batch_group_;
+  auto& merge_paint = rrect_group.paints.front();
+  auto& paint = batch_group_.paints.front();
+
+  if (paint.GetShader() != merge_paint.GetShader()) {
```

Comment:
I don’t think the current batching logic for gradients is correct. I suggest that this change only batches solid colors for now and avoids introducing too much complexity.